### PR TITLE
Merge pull request #75946 from davidtwco/computecpp-init

### PR DESCRIPTION
**Pull request by davidtwco**
###### Motivation for this change
Adds package for [ComputeCpp](https://www.codeplay.com/products/computesuite/computecpp).

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc me!

### CODE REVIEW

File: pkgs/development/compilers/computecpp/default.nix [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+      "ld.ldd"
+      "ldd"
+      "ldd-link"
+      "llvm-spirv"
+      "module-pack-wrapper"
+      "sycl-bundler"
+    ];
+    libaries = stdenv.lib.makeLibraryPath [
+      stdenv.cc.cc.lib
+      ncurses5
```
Comment (davidtwco) [inline pkgs/development/compilers/computecpp/default.nix]: I'd appreciate advice from reviewers on how best to fix a issue with this package. When running the binary, I get the following output:

```
/nix/store/anf6w9ad9577d121w8v2jxhp9m09ycil-computecpp-1.2.0/bin/compute: /nix/store/7fnhs9lg64fgzx13d2skjrckiic2yqsk-ncurses-6.1-20190112-abi5-compat/lib/libtinfo.so.5: no version information available (required by /nix/store/anf6w9ad9577d121w8v2jxhp9m09ycil-computecpp-1.2.0/bin/compute)
```

I believe this is due to the original binary being compiled with an older version of `libtinfo.so.5`, but I'm not sure what I can do about that.

Comment (markuskowa) [reply pkgs/development/compilers/computecpp/default.nix]: My guess is that `libtinfo.so` has been stripped of debug symbols. Does this prevent to program from starting?

**Review by markuskowa — CHANGES_REQUESTED** [review]

File: pkgs/development/compilers/computecpp/default.nix (hunk 1 above)
Comment (davidtwco) [reply pkgs/development/compilers/computecpp/default.nix]: It doesn't, it appears to work fine but just prints that message when invoked.

File: pkgs/development/compilers/computecpp/default.nix [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      ocl-icd
+      zlib
+    ];
+  in
+    ''
+      runHook prePatch
+
```
[the remark this answers is not in the document]
Comment (davidtwco) [reply pkgs/development/compilers/computecpp/default.nix]: Switched to using `autoPatchelfHook`.

**Review by markuskowa — APPROVED** [review]
The derivation looks good to me now but I am no expert in compiler integration. Maybe @bhipple @Ericson2314 @Mic92 @FRidh want to have a look?

**Comment by Mic92** [comment]
Right now it does not find the glibc/stdlibc++ headers:

```
[nix-shell:~/.cache/nixpkgs-review/pr-75946]$ cat main.cpp
#include <iostream>

int main() {
  std::cout << "Hello world" << std::endl;
  return 0;
}

[nix-shell:~/.cache/nixpkgs-review/pr-75946]$ cat main.c
#include <stdio.h>

int main() {
  printf("Hello world\n");
  return 0;
}
[nix-shell:~/.cache/nixpkgs-review/pr-75946]$ compute++ main.cpp -o main
/nix/store/ymhg9r0xambgxpzj0gh6l6lcnk69ykdq-computecpp-1.2.0/bin/compute++: /nix/store/cbl3mpw1hr0b6038j5m9mp9hi38iyjjx-ncurses-6.1-20190112-abi5-compat/lib/libtinfo.so.5: no version information available (required by /nix/store/ymhg9r0xambgxpzj0gh6l6lcnk69ykdq-computecpp-1.2.0/bin/compute++)
/nix/store/ymhg9r0xambgxpzj0gh6l6lcnk69ykdq-computecpp-1.2.0/bin/compute++: /nix/store/cbl3mpw1hr0b6038j5m9mp9hi38iyjjx-ncurses-6.1-20190112-abi5-compat/lib/libtinfo.so.5: no version information available (required by /nix/store/ymhg9r0xambgxpzj0gh6l6lcnk69ykdq-computecpp-1.2.0/bin/compute++)
main.cpp:1:10: fatal error: 'iostream' file not found
#include <iostream>
         ^~~~~~~~~~
1 error generated.

[nix-shell:~/.cache/nixpkgs-review/pr-75946]$ compute main.c -o main
/nix/store/ymhg9r0xambgxpzj0gh6l6lcnk69ykdq-computecpp-1.2.0/bin/compute: /nix/store/cbl3mpw1hr0b6038j5m9mp9hi38iyjjx-ncurses-6.1-20190112-abi5-compat/lib/libtinfo.so.5: no version information available (required by /nix/store/ymhg9r0xambgxpzj0gh6l6lcnk69ykdq-computecpp-1.2.0/bin/compute)
/nix/store/ymhg9r0xambgxpzj0gh6l6lcnk69ykdq-computecpp-1.2.0/bin/compute: /nix/store/cbl3mpw1hr0b6038j5m9mp9hi38iyjjx-ncurses-6.1-20190112-abi5-compat/lib/libtinfo.so.5: no version information available (required by /nix/store/ymhg9r0xambgxpzj0gh6l6lcnk69ykdq-computecpp-1.2.0/bin/compute)
In file included from main.c:1:
/nix/store/4v4ginzfvpsrnvafl1cxx1jn4wy6kc1r-glibc-2.27-dev/include/stdio.h:33:10: fatal error: 'stddef.h' file not found
#include <stddef.h>
         ^~~~~~~~~~
1 error generated.
```

**Comment by davidtwco** [comment]
> 
> 
> Right now it does not find the gcc/stdlibc++ headers:

I'll look into fixing this, thanks.

**Comment by davidtwco** [comment]
@Mic92 I've updated the commit to fix that issue.

**Pull request merged**